snuba/clickhouse/translators/snuba/allowed.py

```diff
@@ -0,0 +1,92 @@
+from typing import TypeVar, Union
+
+from snuba.clickhouse.translators.snuba import SnubaClickhouseStrictTranslator
+from snuba.datasets.plans.translator.mapper import ExpressionMapper
+from snuba.query.expressions import (
+    Argument,
+    Column,
+    CurriedFunctionCall,
+    FunctionCall,
+    Lambda,
+    Literal,
+    SubscriptableReference,
+)
+
+TExpIn = TypeVar("TExpIn")
+TExpOut = TypeVar("TExpOut")
+
+
+class SnubaClickhouseMapper(
+    ExpressionMapper[TExpIn, TExpOut, SnubaClickhouseStrictTranslator]
+):
+    pass
+
+
+# Below are the abstract classes that need to be extended in order to
+# write a mapping rule SnubaClickhouseMappingTranslator would accept.
+# Nothing broader that does not fit in these types constraints is
+# allowed.
+
+
+class LiteralMapper(SnubaClickhouseMapper[Literal, Literal]):
+    """
+    Literals can only be translated into Literals.
+    """
+
+    pass
+
+
+class ColumnMapper(SnubaClickhouseMapper[Column, Union[Column, Literal, FunctionCall]]):
+    """
+    Columns can be translated into other Columns, or Literals (if a Snuba
+    column has a hardcoded value on a storage like the event type on
+    transactions), or FunctionCalls (common when wrapping Columns into
+    assertNotNull calls).
+    """
+
+    pass
+
+
+class FunctionCallMapper(SnubaClickhouseMapper[FunctionCall, FunctionCall]):
+    """
+    Functions are only allowed to become Functions so that we can ensure
+    CurriedFunctions internal functions can be successfully translated.
+
+    TODO: We actually need to loosen this constraint (for functions
+    translating into pre aggregated Column). Though before doing so we
+    need to give a dedicated type that will be stricter for the
+    CurriedFunction internal function.
```

Comment:
FunctionCallMapper here only allows to translate a function into a function.
Any translation we may want to do for a multi table storage (where one of the tables is actually pre-aggregated) would require us to translate FunctionCall into Column. Since we will receive a function (count()) and we will have to translate into a column that has the aggregation.

If I allow this here, that would ruin all the effort to keep translation type safe for CurreidFunctionCall since the inner function cannot be a Column. Though as soon as I swap the inner function with its own type (see comment above) this will not be a problem anymore.

Reply:
I just saw this after looking at https://github.com/getsentry/snuba/pull/936#discussion_r430861407. This makes more sense now.